feat: campaign monitor implementation

dkoo opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

Adds Campaign Monitor as an email service provider (ESP) option, in addition to Mailchimp and Constant Contact.

The UI remains mostly the same as the other ESPs, except for the selection of Lists vs. Segments when selecting whom to send a campaign to. Campaign Monitor allows campaigns to be sent to either lists or segments.

The back-end for this ESP differs slightly from the other ESPs. Campaign Monitor's API does not support updating of existing campaigns, so instead of creating a campaign on newsletter post creation and keeping it in sync upon saving the post, the Campaign Monitor class stores required campaign data as post meta. Whenever the user sends a test email or publishes the campaign, the plugin will generate a new campaign and send it immediately. If a test campaign, the campaign will be silently cleaned up on successful test send, to avoid cluttering the Campaign Monitor dashboard with duplicate campaigns.

In terms of validation, campaign data will be validated on the fly as it is updated in the Gutenberg editor. The "send" button is only enabled if the post contains all data required to send a campaign.

Lastly, when sending a campaign, the plugin will now automatically save the post before sending. This is to ensure that any dirty/unsaved data present in the editor is saved before the campaign is generated and sent. This shouldn't affect how Mailchimp and Constant Contact work, since those ESPs sync their campaigns upon saving and sending anyway, but it will prevent errors with Campaign Monitor sending campaigns with the wrong data if the edited post and saved post differ.

#### To do

Campaign Monitor supports sending to multiple lists or segments simultaneously. Should we consider updating the UI to a multiselect so editors can select more than one at a time? cc @thomasguillot 

#### A security question

The [Campaign Monitor API](https://www.campaignmonitor.com/api/campaigns/#creating-draft-campaign) ingests campaign content in the form of a public URL that serves the HTML content of the email. This means that in order to create campaigns, we need to provide a public REST API endpoint that displays the rendered MJML content for a newsletter, even when still in draft status. Right now the endpoint [is registered as public](https://github.com/Automattic/newspack-newsletters/pull/364/files#diff-da59c086ec29c628c0623601aca945fba831d2efda222ca4a5a0fbb602a195abR49) and doesn't have any permission checks. Is there any way we can think of to further lock this down in case publishers are concerned about the security of this? cc @jeffersonrabb and/or @adekbadek 

### How to test the changes in this Pull Request:

#### Setup

1. Create a Campaign Monitor account if you don't already have one: https://www.campaignmonitor.com/
2. Go to **Account Settings > API Keys**. Generate an API key if your account doesn't have one. Copy both the API Key and Client ID.
3. Ensure that your account has at least two subscriber lists, and that at least one of those lists has at least a segment or two (create them if necessary). See here for info on creating [lists](https://help.campaignmonitor.com/create-a-subscriber-list) and [segments](https://help.campaignmonitor.com/list-segmentation).
4. Check out this branch and run `npm run build`.
5. In WP Admin, go to **Newsletters > Settings** and note a new "Campaign Monitor" option in the "Service Provider" dropdown, as well as new settings fields for Campaign Monitor API Key and Campaign Monitor Client ID.
6. Select Campaign Monitor as your Service Provider but leave the Campaign Monitor fields blank. Click "save changes" to save.

#### Creating a new campaign

1. In WP Admin, go to **Newsletters > Add New**.
2. Confirm that you see the modal to enter your API credentials. Enter your API Key and Client ID and click "save settings" to save.
3. Select a layout and edit the campaign content as usual.
4. Note the **Newsletter** panel in the Document sidebar. It should have fields relevant for Campaign Monitor:

<img width="267" alt="Screen Shot 2020-10-23 at 4 10 47 PM" src="https://user-images.githubusercontent.com/2230142/97058190-57384200-154a-11eb-86b5-c16c3dd8a898.png">

5. Under **Send Mode**, select the option for "List." Confirm that a select dropdown appears with options that match the lists you've created in your Campaign Monitor account. Also confirm that a link appears to a campaignmonitor.com help document on creating lists.
6. Switch **Send Mode** to "Segment." Confirm that the dropdown changes to show the segments you've created in Campaign Monitor. Also confirm that a link appears to a campaignmonitor.com help document on creating segments.
7. Decide whether you want to send to a list or segment to test. Select the relevant **Send Mode** option, then select a list or segment in the dropdown.
8. Fill in the rest of the required campaign info (Subject, From name/email). Confirm that once all fields are filled out, the "Send" button at the top of the editor becomes available/enabled. (Also confirm that if you clear any of these fields, it becomes disabled again.)

#### Sending a test campaign

1. In the editor sidebar, expand the **Testing** panel.
2. Enter one or more email addresses and click the "Send a test email" button.
3. Confirm that the campaign gets delivered to the email addressees entered.
4. To test campaign validation and error handling, clear out one of the required fields (Subject, Send Mode, List or Segment, From name/email) and click the "Send a test email" button again to attempt another test email. This time, the editor should throw an error notice stating that certain campaign info is missing.
5. To test the temporary campaign cleanup, go to your Campaign Monitor dashboard and go to Campaigns. Confirm that there are no extra campaigns in the Overview, Drafts, or Trash tabs from your test sends.

#### Sending a live campaign

1. Ensure that all required campaign info fields are filled in (Subject, Send Mode, List or Segment, From name/email).
2. Confirm that the "Send" button is enabled. Click it and confirm that the post saves automatically before showing the presend modal.
3. The presend modal should look something like this:

<img width="360" alt="Screen Shot 2020-10-23 at 4 20 19 PM" src="https://user-images.githubusercontent.com/2230142/97058665-ad59b500-154b-11eb-8b77-a3ea9bc836ae.png">

4. Click "send" to send the campaign. Confirm that a.) the campaign gets delivered to all subscriber addresses in the list or segment you selected, and b.) the "campaign has been sent" success message isn't shown until after the campaign is published and delivered.
5. To test error handling on send, manually throw an exception in the `send` function in `class-newspack-newsletters-campaign-monitor.php` after line ~329. e.g. `throw new Exception( __( 'A manual send error.', 'newspack-newsletters' ) );`
6. Create a new newsletter and fill in the required campaign info, then click "send".
7. This time, confirm that the editor shows an error notice with your exception's error message, and the post does not transition to "published" status. The "campaign sent" success message should not be shown.

#### Scheduling a live campaign

1. Create a new newsletter and fill out the content and all required fields.
2. In the Document sidebar, change the **Publish: Immediately** option to a time in the near future, e.g. 10 minutes from now.
3. Click "Schedule sending" to schedule the campaign.
4. Confirm that the campaign gets delivered at the time you specified.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

dkoo commented:
Keeping this as draft for now since I still want to run some tests to ensure that Mailchimp and CC aren't affected by these changes.

adekbadek reviewed (changes requested):
Some other issues:

1. When trying to send a test campaign, I got the `Campaign Monitor error: Missing campaign sender data.` error – I first had to save draft to get the metadata saved. This is not terrible, but just flagging here. 
2. Retrieving data from the API might take a moment, during which the list selector is disabled – how about a spinner here?

![image](https://user-images.githubusercontent.com/7383192/97866755-d2e57c00-1d0c-11eb-81df-4aeae15b9f7c.png)

3. When the pre-send modal appears, the button is not disabled despite data being still loaded:

![image](https://user-images.githubusercontent.com/7383192/97866794-edb7f080-1d0c-11eb-9014-74ca345ca58a.png)

dkoo commented on src/newsletter-editor/editor/index.js:
```diff
@@ -33,16 +29,8 @@ const Editor = compose( [
-			'color',
-			'palette',
-		] );
-		const colors = settings.colors || experimentalSettingsColors || [];
```
Whoops, sorry—fixed in f09dd41.

dkoo commented on src/components/send-button/index.js:
```diff
@@ -67,12 +67,15 @@ export default compose( [
 
+		const hasCampaign =
+			newsletterData.campaign ||
+			window.newspack_newsletters_data.service_provider === 'campaign_monitor';
```
Good suggestion—done in c6e1df0.

dkoo commented:
> When trying to send a test campaign, I got the `Campaign Monitor error: Missing campaign sender data.` error – I first had to save draft to get the metadata saved. This is not terrible, but just flagging here.

When I try to replicate, the behavior is slightly different than you describe—clicking "Send a test email" does save the post automatically. If I create a new newsletter, fill out all of those fields and click the "Send a test email" button before sending, the test email gets sent successfully. But if any of the Send Mode, List/Segment, To, or From Name/Email fields are blank, that's what throws the `Missing campaign sender data` error. This is unfortunately unavoidable because the Campaign Monitor API requires all of these fields in order to create a campaign (even just to send a test).

> how about a spinner here?

Sure! Added in 6a1c672 and allowed interactivity with the dropdowns even if empty, as long as they're not in a loading state, in 3b00157.

> When the pre-send modal appears, the button is not disabled despite data being still loaded:


Fixed in c74c916.

dkoo commented:
@adekbadek I'm not sure why this started happening all of a sudden, but I'm getting the following error from CircleCI now:

```bash
#!/bin/bash -eo pipefail
composer install
composer global require "phpunit/phpunit=5.7.*"
rm -rf $WP_TESTS_DIR $WP_CORE_DIR
bash bin/install-wp-tests.sh wordpress_test root '' 127.0.0.1 latest
phpunit

Installing dependencies from lock file (including require-dev)
Verifying lock file contents can be installed on current platform.
Your lock file does not contain a compatible set of packages. Please run composer update.

  Problem 1
    - dealerdirect/phpcodesniffer-composer-installer is locked to version v0.6.2 and an update of this package was not requested.
    - dealerdirect/phpcodesniffer-composer-installer v0.6.2 requires composer-plugin-api ^1.0 -> found composer-plugin-api[2.0.0] but it does not match the constraint.

You are using Composer 2, which some of your plugins seem to be incompatible with. Make sure you update your plugins or report a plugin-issue to ask them to support Composer 2.

Exited with code exit status 2
CircleCI received exit code 2
```

Any idea how to resolve? I ran `composer update` but this doesn't seem to fix the problem. Do we need to update that `dealerdirect/phpcodesniffer-composer-installer` version in `composer.json`, maybe?

adekbadek commented:
> Any idea how to resolve? I ran composer update but this doesn't seem to fix the problem. Do we need to update that dealerdirect/phpcodesniffer-composer-installer version in composer.json, maybe?

Yeah, same thing happened in Campaigns – https://github.com/Automattic/newspack-popups/pull/275/files
Just `composer remove dealerdirect/phpcodesniffer-composer-installer && composer require  --dev dealerdirect/phpcodesniffer-composer-installer`

adekbadek commented:
> If I create a new newsletter, fill out all of those fields, add an email to receive a test email and click the "Send a test email" button before saving, the test email gets sent successfully.

This is odd – on a local VVV instance the newsletter is _not saved_ the first time I click "Send Test Email". But on an ephemeral test site it works as expected, so let's forget it.

adekbadek reviewed (changes requested):
I've updated the composer configuration expecting the error to be same as fixed in https://github.com/Automattic/newspack-popups/pull/275/files, but that was not enough. Encountering a [`phpunit: command not found` error](https://app.circleci.com/pipelines/github/Automattic/newspack-newsletters/876/workflows/34812231-4bf7-4353-803b-e8190fa20541/jobs/1756) I've changed the CI config to use a local `phpunit` in 773ba385134ac9ac89ffcfebd757d5ddb80ffd89 – that did the trick.

As for the code review, I've left one comment.

dkoo commented on src/components/send-button/index.js:
```diff
@@ -195,7 +198,11 @@ export default compose( [
-							disabled={ newsletterValidationErrors.length > 0 }
+							disabled={
+								newsletterValidationErrors.length > 0 ||
+								( window.newspack_newsletters_data.service_provider === 'campaign_monitor' &&
```
I now remember why I didn't originally do what you suggested in c6e1df0—for some reason, after saving, `newsletterData` briefly becomes empty during the exact time we need to read it for pre-send checks. You can verify by adding `console.log( newsletterData )` to `send-button/index.js` and watching the console as you click the "Send" button.

This causes the pre-send validation to not work for Campaign Monitor. Also, because the Campaign Monitor validation happens a bit differently (based on the presence of metadata, rather than synced campaign data), the same checks for `newsletterValidationErrors` aren't needed here.

This works for me, even though it's not quite as clean as keeping everything in the ESP files: 081865d

Happy to refactor if you can help me figure out why `newsletterData` keeps disappearing?

adekbadek replied:
The disappearance happened because meta fields in the editor were updated on post saving. Since these two meta fields were not registered, they were missing in the response. 
I've registered them in cb4134207a276b8f01b9193795560f479aeaba2b
Let me know if that solves the validation issue.

Also, I noticed that all the meta fields are public – retrievable via the REST API. To prevent that, I've also updated all existing meta fields to be be available only in the editor.

dkoo replied:
This makes a lot of sense to me! Thanks for explaining. I did notice that the "send" button became disabled after saving because the the `newsletterData.campaign` value kept getting lost. I think d35c662 fixes it.

dkoo commented:
@adekbadek Thanks for fixing the composer stuff!

dkoo commented:
> This is odd – on a local VVV instance the newsletter is not saved the first time I click "Send Test Email". But on an ephemeral test site it works as expected, so let's forget it.

Strange, it also saves for me in master (I didn't change anything with the test stuff): https://github.com/Automattic/newspack-newsletters/blob/master/src/newsletter-editor/testing/index.js#L37

dkoo commented:
@adekbadek Is this okay to merge now, or do we need to pull in another reviewer to approve your changes?

adekbadek reviewed (approved):
:shipit:
